controllers/machine_controller_node_labels.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	apicorev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	"sigs.k8s.io/cluster-api/controllers/external"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/patch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (r *MachineReconciler) reconcileInterruptibleNodeLabel(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) (ctrl.Result, error) {
+	logger := ctrl.LoggerFrom(ctx)
+
+	// Check that the Machine hasn't been deleted or in the process
+	// and that the Machine has a NodeRef.
+	if !machine.DeletionTimestamp.IsZero() || machine.Status.NodeRef == nil {
+		return ctrl.Result{}, nil
+	}
+
+	// Get the infrastructure object
+	infra, err := external.Get(ctx, r.Client, &machine.Spec.InfrastructureRef, machine.Namespace)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Get interruptible instance status from the infrastructure provider.
+	interruptible, _, err := unstructured.NestedBool(infra.Object, "status", "interruptible")
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get interruptible status from infrastructure provider for Machine %q in namespace %q: %w", machine.Name, machine.Namespace, err)
```

Comment:
@vincepri what do you think about logging here and returning early instead of returning an error? If we return an error, it means we will start exponential backoff retries. I assume that unless the infra machine changes, continuing to retry the same object will yield the same failure each time. And, given we are already watching for infra machine changes, the machine will get requeued whenever the infra machine is updated.

Reply:
We definitely can. That said, these errors are uncommon code paths where we shouldn't get into unless something really wrong happened. Most of the time, functions like NestedBool are going to return regardless if the value exists or not, and in this case we use the default value for the boolean if the json path (status.interruptible) doesn't exists. The only way for this call to fail is if the `infra.Object` isn't well formed, which is unlikely in most cases, and it would probably hint to data corruption

Reply:
Exactly, and if it's malformed data, exponential backoff isn't going to be useful here.

Reply:
That makes sense, thank you.